docs/studies/hidden_linear_function.ipynb

```diff
@@ -0,0 +1,540 @@
+    "Consider $A \\in \\mathbb{F}_2^{n \\times n}$ - upper-triangular binary matrix of size $n \\times n$, $b \\in \\mathbb{F}_2^n$ - binary vector of length $n$.\n",
+    "\n",
+    "Define a function $q : \\mathbb{F}_2^n \\to \\mathbb{Z}_4 $:\n",
+    "\n",
+    "$$q(x) = (2 x^T A x + b^T x) ~\\text{mod}~ 4, $$ \n",
+    "\n",
+    "Also define\n",
+    "\n",
+    "$$\\mathcal{L}_q = \\Big\\{x \\in  \\mathbb{F}_2^n : q(x \\oplus y) = (q(x) + q(y)) ~\\text{mod}~ 4 ~~ \\forall y \\in \\mathbb{F}_2^n \\Big\\}.$$\n",
+    "\n",
+    "Turns out (see Lemma 1 on page 6 in [1]) that restriction of $q$ on $\\mathcal{L}_q$ is a linear function, i.e. there exists such $z \\in \\mathbb{F}_2^n$, that\n",
+    "\n",
+    "$$q(x) = 2 z^T x ~~\\forall x \\in \\mathcal{L}_q.$$\n",
+    "\n",
+    "Our task is, given $A$ and $b$, to find $z$. There may be multiple answers - we need to find any such answer.\n",
+    "\n",
+    "### Why is this problem interesting?\n",
+    "\n",
+    "#### 1. It's a problem without an oracle\n",
+    "\n",
+    "There are other problems where task is to find coefficients of a linear function. But usually the linear function is represented by an [oracle](https://en.wikipedia.org/wiki/Oracle_machine). See, for example, [Bernstein–Vazirani algorithm](https://en.wikipedia.org/wiki/Bernstein%E2%80%93Vazirani_algorithm).\n",
+    "\n",
+    "In this problem we avoid use of an oracle by \"hiding\" the linear function in the coefficients of quadratic form $A$ and $b$, which are the only inputs to the problem.\n",
+    "\n",
+    "#### 2. Quantum circuits have advantage over classical when solving this problem\n",
+    "\n",
+    "As we will show below, this problem can be solved with a Clifford circuit. Therefore, according to the [Gottesman–Knill theorem](https://en.wikipedia.org/wiki/Gottesman%E2%80%93Knill_theorem), this problem can be solved in polynomial time on a classical computer. So, it might look like quantum computers aren't better than classical ones in solving this problem.\n",
+    "\n",
+    "However, if we apply certain restrictions on matrix $A$, the circuit will have fixed depth (i.e. number of `Moment`s). Namely, if the matrix $A$ is an adjacency matrix of a \"grid\" graph (whose edges can be colored in 4 colors), all CZ gates will fit in 4 `Moment`s, and overall we will have only 8 `Moment`s - and this doesn't depend on $n$.\n",
+    "\n",
+    "But for classical circuits it can be proven (see [1]) that even if we restrict matrix $A$ in the same way, the depth of classical circuit (with gates of bounded fan-in) must grow as $n$ grows (in fact, it grows as $\\log(n)$).\n",
+    "\n",
+    "In terms of complexity theory, this problem is in [QNC<sup>0</sup>](https://complexityzoo.uwaterloo.ca/Complexity_Zoo:Q#qnc0), but not in [NC<sup>0</sup>](https://en.wikipedia.org/wiki/NC_(complexity)), which shows that\n",
+    "QNC<sup>0</sup> $\\nsubseteq$ NC<sup>0</sup>.\n",
+    "\n",
+    "### Structure of this notebook\n",
```

Comment:
This section seems out of place.  Should it be part of the introduction?

Reply:
This was part of the introduction. I moved the problem statement in the separate chapter, so the introduction is shorter. Is it better now?